teerex/src/lib.rs

```diff
@@ -162,70 +169,143 @@ pub mod pallet {
+					<MrEnclave>::decode(&mut proof.as_slice()).unwrap_or_default(),
+					MrSigner::default(),
+					<timestamp::Pallet<T>>::get().saturated_into(),
+					SgxBuildMode::default(),
+					SgxStatus::Invalid,
+				)
+				.with_pubkey(sender.encode().as_ref())
+				.with_attestation_method(SgxAttestationMethod::Skip { proxied }),
+			};
 
-			#[cfg(not(feature = "skip-ias-check"))]
 			if !<SgxAllowDebugMode<T>>::get() && enclave.build_mode == SgxBuildMode::Debug {
 				log::warn!("teerex: debug mode is not allowed to attest!");
 				return Err(<Error<T>>::SgxModeNotAllowed.into())
 			}
 
-			#[cfg(feature = "skip-ias-check")]
-			log::warn!("[teerex]: Skipping remote attestation check. Only dev-chains are allowed to do this!");
-
-			#[cfg(feature = "skip-ias-check")]
-			let enclave = SgxEnclave::<Vec<u8>>::new(
-				SgxReportData::default(),
-				// insert mrenclave if the ra_report represents one, otherwise insert default
-				<MrEnclave>::decode(&mut ra_report.as_slice()).unwrap_or_default(),
-				MrSigner::default(),
-				<timestamp::Pallet<T>>::get().saturated_into(),
-				SgxBuildMode::default(),
-				SgxStatus::Invalid,
-			)
-			.with_pubkey(&sender.encode())
-			.with_url(worker_url.clone());
-
-			Self::add_enclave(&sender, &enclave)?;
-
-			#[cfg(not(feature = "skip-ias-check"))]
+			let enclave = match worker_url {
+				Some(ref url) => enclave.with_url(url.clone()),
+				None => enclave,
+			};
+
+			Self::add_enclave(&sender, &MultiEnclave::from(enclave.clone()))?;
+			Self::poke_shard(enclave.mr_enclave.into(), &sender)?;
+
 			Self::deposit_event(Event::AddedEnclave {
 				registered_by: sender,
 				worker_url,
 				tcb_status: Some(enclave.status),
 				attestation_method: enclave.attestation_method,
 			});
-
-			#[cfg(feature = "skip-ias-check")]
-			Self::deposit_event(Event::AddedEnclave {
-				registered_by: sender,
-				worker_url,
-				tcb_status: None,
-				attestation_method: SgxAttestationMethod::Skip { proxied: false },
-			});
 			Ok(().into())
 		}
 
 		#[pallet::call_index(1)]
 		#[pallet::weight((<T as Config>::WeightInfo::unregister_enclave(), DispatchClass::Normal, Pays::Yes))]
-		pub fn unregister_enclave(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-			log::info!("teerex: called into runtime call unregister_enclave()");
+		pub fn unregister_sovereign_enclave(
+			origin: OriginFor<T>,
+			enclave_signer: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			log::info!("teerex: called into runtime call unregister_sovereign_enclave()");
 			let sender = ensure_signed(origin)?;
-
-			Self::remove_enclave(&sender)?;
-			Self::deposit_event(Event::RemovedEnclave(sender));
+			let enclave = Self::sovereign_enclaves(&enclave_signer)
+				.ok_or(<Error<T>>::EnclaveIsNotRegistered)?;
+			let now = <timestamp::Module<T>>::get();
+			let oldest_acceptable_attestation_time =
+				now.saturating_sub(T::MaxSilenceTime::get()).saturated_into::<u64>();
```

Comment:
I think we should rename this to `T::AttestationTimeout` or something like this.

Reply:
max silence time also renews upon activity. not only attestation. We could use MaxPassiveTime?

Reply:
not much better IMO

Reply:
I also thought that this is the design, but when I looked at the code, I could only find that it is renewed upon a new attestation. But in this case, I guess I was wrong, and then it is ok as it is